Add an error callback.

robholland opened the pull request:
This is useful to report exceptions to an exception monitoring service.

robholland commented:
I stubbed out all of the history/state machine innards as setting up fabricators fully enough for all that code to function was complicated and didn't seem to be worth it for the contexts I wanted to test.

antstorm commented on lib/temporal/activity/poller.rb:
```diff
@@ -66,6 +66,11 @@ def poll_for_task
       rescue StandardError => error
         Temporal.logger.error("Unable to poll activity task queue: #{error.inspect}")
+
+        Temporal.configuration.error_handlers.each do |handler|
```
Let's get this under a more concise interface, something like `Temporal::ErrorHandler.handle(error, nil)`

antstorm commented on lib/temporal/activity/poller.rb:
```diff
@@ -66,6 +66,11 @@ def poll_for_task
         Temporal.logger.error("Unable to poll activity task queue: #{error.inspect}")
+
+        Temporal.configuration.error_handlers.each do |handler|
+          handler.call(error, nil)
```
We'll also need to put guards around call to the provided lambdas, we can't afford crashing on buggy error handling

antstorm commented on lib/temporal/configuration.rb:
```diff
@@ -3,7 +3,7 @@
 module Temporal
   class Configuration
-    attr_reader :timeouts
+    attr_reader :timeouts, :error_handlers
```
Let's add a method for adding these handlers rather than expose the whole array. Something like:

```ruby
def on_error(&block)
  @error_handlers << block
end
```

antstorm commented on lib/temporal/workflow/poller.rb:
```diff
@@ -60,6 +60,11 @@ def poll_for_task
       rescue StandardError => error
         Temporal.logger.error("Unable to poll workflow task queue: #{error.inspect}")
+
+        Temporal.configuration.error_handlers.each do |handler|
```
Not sure about this one… or maybe we need to have multiple handler types

robholland replied:
We need to be able to report WorkflowNotRegistered.

robholland replied:
This is required to catch things like an undefined workflow class. I don't really see a reason to treat this any differently. What's your concern here?

antstorm commented on lib/temporal/workflow/poller.rb:
```diff
@@ -60,6 +60,9 @@ def poll_for_task
       rescue StandardError => error
         Temporal.logger.error("Unable to poll workflow task queue: #{error.inspect}")
+
+        Temporal::ErrorHandler.handle(error, nil)
```
I think this is expected to fail on a polling timeout, so I don't think it's super helpful to report. Maybe we need to rescue different errors here to distinguish expected failure

robholland replied:
If that were the case then we'd see "Unable to poll workflow task queue" all over the logs, but we don't. So I think failure here is unexpected.

antstorm replied:
That's a good point, you're probably right. Please ignore it then

antstorm commented on lib/temporal/workflow.rb:
```diff
@@ -19,6 +19,8 @@ def self.execute_in_context(context, input)
       Temporal.logger.error("Workflow execution failed with: #{error.inspect}")
       Temporal.logger.debug(error.backtrace.join("\n"))
 
+      Temporal::ErrorHandler.handle(error, context.send(:metadata))
```
We should probably make `metadata` public, `.send` suggests that we're doing something wrong here

robholland replied:
Done.

antstorm reviewed (commented):
Any ideas why the specs are failing? Looks relevant

robholland commented:
Fixed the example specs failing, due to local workflow context not supporting metadata. I've added that now and adjusted callers to pass that rather than plain headers.

antstorm commented on lib/temporal/activity/task_processor.rb:
```diff
@@ -26,16 +29,17 @@ def process
         # Do not complete asynchronous activities, these should be completed manually
         respond_completed(result) unless context.async?
       rescue StandardError, ScriptError => error
+        Temporal::ErrorHandler.handle(error, metadata: metadata.to_h)
```
Btw, I see where you coming from with `.to_h`, but compared to logs it's actually that bad to pass the whole metadata object here. Any reasons you decided not to do it?

antstorm replied:
Is would allow devs to distinguish between different metadata types

robholland replied:
I didn't want application developers to have to know the metadata class internals. If we send a hash then there is no API concerns. I don't feel strongly about it though if you'd prefer to send the objects I can `.to_h` in my application's handler.

robholland commented on lib/temporal/metadata/activity.rb:
```diff
@@ -23,6 +23,21 @@ def initialize(namespace:, id:, name:, task_token:, attempt:, workflow_run_id:,
+      def to_h
+        {
+          'attempt' => attempt,
+          'headers' => headers,
```
Ok, I confess I don't really know what they are used for, I'll remove.

robholland commented on lib/temporal/metadata/activity.rb:
```diff
@@ -23,6 +23,21 @@ def initialize(namespace:, id:, name:, task_token:, attempt:, workflow_run_id:,
+          'activity_id' => id,
+          'activity_name' => name,
+          'namespace' => namespace,
+          'task_token' => task_token,
```
We used it to debug a Temporal/Ruby-SDK bug, not sure how useful it would be in general.

robholland commented on lib/temporal/metadata/activity.rb:
```diff
@@ -23,6 +23,21 @@ def initialize(namespace:, id:, name:, task_token:, attempt:, workflow_run_id:,
+        {
+          'attempt' => attempt,
+          'headers' => headers,
+          'heartbeat_details' => heartbeat_details,
```
This is important to send, the concept of heartbeat details is to provide cursors for long running activities. If you don't have the heartbeat details alongside an exception you can know which slice/chunk/batch it was working on when it blew up.

robholland commented on lib/temporal/activity/task_processor.rb:
```diff
@@ -6,9 +6,12 @@
 module Temporal
   class Activity
     class TaskProcessor
+      attr_accessor :metadata
```
Right, doesn't need to be, will fix.

robholland commented on lib/temporal/activity/task_processor.rb:
```diff
@@ -26,16 +29,17 @@ def process
         context = Activity::Context.new(client, metadata)
 
-        result = middleware_chain.invoke(metadata) do
+        result = middleware_chain.invoke(metadata, task.to_h) do
```
If I had been able to see the task alongside the exception we saw then it would have been immediately obvious what the issue was (query being sent with no workflow tasks). Maybe I'm bias as an SDK contributor, but this information is extremely valuable to me.

antstorm commented:
@robholland can you please rebase it and I'll get it merged?

robholland commented:
Rebased.

antstorm commented:
Thank you, @robholland , great PR!
